op-acceptance-tests/tests/base/reth_geth_divergence_test.go

```diff
@@ -0,0 +1,190 @@
+package base
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/devnet-sdk/system"
+	"github.com/ethereum-optimism/optimism/devnet-sdk/testing/systest"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+)
+
+// TestRethGethDivergence checks that reth and geth do not diverge in terms of state root.
+// geth is considered the canonical/correct implementation.
+func TestRethGethDivergence(t *testing.T) {
+	systest.SystemTest(t, rethGethDivergenceTestScenario())
+}
+
+func rethGethDivergenceTestScenario() systest.SystemTestFunc {
+	const (
+		// testTimeout is the maximum time allowed for the test to run
+		testTimeout = 60 * time.Second
+	)
+
+	return func(t systest.T, sys system.System) {
+		logger := testlog.Logger(t, log.LevelInfo)
+		logger.Info("Started reth-geth divergence test")
+
+		// Check all L2 chains
+		for i, chain := range sys.L2s() {
+			chainIndex := i
+			currentChain := chain
+			t.Run(fmt.Sprintf("Chain_%d", chainIndex), func(t systest.T) {
+				t.Parallel()
+				checkChainForDivergence(t, currentChain, logger.New("chain", chainIndex), testTimeout)
+			})
+		}
+	}
+}
+
+// checkChainForDivergence checks a single chain for reth/geth divergence
+func checkChainForDivergence(
+	t systest.T,
+	chain system.L2Chain,
+	logger log.Logger,
+	testTimeout time.Duration,
+) {
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(t.Context(), testTimeout)
+	defer cancel()
+
+	// Find geth and all reth nodes
+	gethNode, rethNodes, err := findGethAndRethNodes(chain, logger)
+	if err != nil {
+		t.Skip(err.Error())
+		return
+	}
+
+	logger.Info("Found geth and reth nodes",
+		"geth", gethNode.Name(),
+		"reth_count", len(rethNodes))
+
+	// Get client for geth node
+	gethClient, err := gethNode.GethClient()
```

Comment:
Just wondering if it would be simpler and more reliable to just form a slice of all nodes (and their clients) across the network and check for divergence across all of them i.e. find the smallest latest block across all of them, and check if its state root matches for the same block number across all the client?

wdyt

Reply:
Sounds good. I've implemented this in my latest push; please take a look.
Thanks again!